File: dsa/arrays,hashing/49. group_anagrams.cpp
/*
LC 49 — Group Anagrams
Pattern: Hash Map + Frequency Count (Counting Sort Key)
Difficulty: Medium
Time: __ min
Approach: Avoid O(k log k) sorting by counting character frequencies. 
          Create a string key like "1#0#2..." where each number represents 
          the count of 'a' through 'z'. Use this unique key in a hash map.
Time Complexity: O(n * k) — where n is number of strings, k is max string length.
Space Complexity: O(n * k)
*/

#include<bits/stdc++.h>
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> mp;
        
        for (const string& s : strs) {
            // Step 1: Count character frequencies (a-z)
            int count[26] = {0};
            for (char c : s) {
                count[c - 'a']++;
            }
            
            // Step 2: Build a unique key based on counts
            // Using '#' ensures counts like (1, 11) don't collide with (11, 1)
            string key = "";
            for (int i = 0; i < 26; i++) {
                key += "#";
                key += to_string(count[i]);
            }
            
            // Step 3: Group by key
            mp[key].push_back(s);
        }
        
        // Step 4: Extract groups from map
        vector<vector<string>> ans;
        for (auto& it : mp) {
            ans.push_back(it.second);
        }
        return ans;
    }
};

File: dsa/arrays,hashing/49. group_anagrams.py
"""
LC 49 — Group Anagrams
Pattern: Hash Map + Frequency Count (Tuple Key)
Difficulty: Medium
Time: __ min
Approach: Use an immutable tuple of 26 character counts as a dictionary key.
          Anagrams have identical counts for every character, so they
          will map to the same tuple key.
Time Complexity: O(n * k) — where n is number of strings, k is max string length.
Space Complexity: O(n * k)
"""
from typing import List
from collections import defaultdict

class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # defaultdict(list) handles missing keys by creating an empty list
        mp = defaultdict(list)

        for s in strs:
            # Step 1: Initialize list of 26 zeros
            count = [0] * 26
            for char in s:
                count[ord(char) - ord('a')] += 1
            
            # Step 2: Convert list to tuple to use as hashable dict key
            # Step 3: Append original string to the group
            mp[tuple(count)].append(s)

        # Step 4: Convert map values view to a final nested list
        return list(mp.values())